hugo/static/cgi-bin/watermark.cgi

```diff
@@ -0,0 +1,166 @@
+    """Resolve location of add_footer.py relative to this CGI script."""
+    here = Path(__file__).resolve()
+    homedir = Path(os.environ.get('HOME', '/home/anthologizer'))
+    # Walk up to locate bin/add_footer.py
+    for parent in [here.parent, *here.parents, homedir / "acl-anthology"]:
+        candidate = parent.parent / "bin" / "add_footer.py" if parent.name == 'cgi-bin' else parent / "bin" / "add_footer.py"
+        if candidate.exists():
+            return candidate
+    # Fallback relative guess (3 levels up)
+    guess = here.parents[2] / 'bin' / 'add_footer.py'
+    return guess
+
+def main():
+    # Debug flag: append ?debug=1 to request URL to get full stderr on failure.
+    qs = os.environ.get('QUERY_STRING', '')
+    qparams = dict(urllib.parse.parse_qsl(qs, keep_blank_values=True)) if qs else {}
+    debug_mode = qparams.get('debug') in {'1','true','yes','on'}
+    method = os.environ.get('REQUEST_METHOD', 'GET').upper()
+    if method != 'POST':
+        http_error(405, 'Use POST with multipart/form-data.')
+
+    try:
+        length = int(os.environ.get('CONTENT_LENGTH', '0'))
+    except ValueError:
+        length = 0
+    if length <= 0:
+        http_error(400, 'Empty request body.')
+    if length > MAX_BYTES:
+        http_error(400, f'File too large (> {MAX_BYTES//1024//1024}MB).')
+
+    form = cgi.FieldStorage()
+
+    if 'pdf' not in form or not getattr(form['pdf'], 'file', None):
+        http_error(400, 'Missing PDF file.')
+    pdf_item = form['pdf']
+    footer_text = form.getfirst('footer_text', '')[:10000]  # cap length
+    page_start_raw = form.getfirst('page_start')
+    page_start = None
+    if page_start_raw:
+        try:
+            page_start = int(page_start_raw)
+            if page_start < 1:
+                raise ValueError
+        except ValueError:
+            http_error(400, 'Invalid page_start (must be positive integer).')
+
+    # Write uploaded PDF to temp file
+    tmp_dir = tempfile.mkdtemp(prefix='wmk_')
+    input_pdf = Path(tmp_dir) / 'input.pdf'
+    output_pdf = Path(tmp_dir) / 'output.pdf'
+
+    with open(input_pdf, 'wb') as f:
+        # stream copy to avoid loading entire file in memory
+        chunked = 0
+        while True:
+            buf = pdf_item.file.read(64 * 1024)
+            if not buf:
+                break
+            chunked += len(buf)
+            if chunked > MAX_BYTES:
+                f.close()
+                http_error(400, 'File exceeded size limit during upload.')
+            f.write(buf)
+
+    # Validate PDF magic
+    try:
+        with open(input_pdf, 'rb') as f:
+            head = f.read(8)
+            if b'%PDF-' not in head:
+                http_error(400, 'Uploaded file is not a PDF.')
+    except Exception:
+        http_error(400, 'Could not read uploaded PDF.')
+
+    add_footer = find_add_footer()
+    if not add_footer.exists():
+        http_error(500, 'Server configuration error: add_footer.py not found.')
```

Comment:
I’d probably move this higher, before doing anything with the uploaded data at all.